pkg/image/admission/imagequalify/qualify.go

```diff
@@ -0,0 +1,46 @@
+package imagequalify
+
+import (
+	"path"
+
+	"github.com/openshift/origin/pkg/image/admission/imagequalify/api"
+	"github.com/openshift/origin/pkg/image/admission/imagequalify/api/validation"
+)
+
+// matchImage attempts to match image against each pattern defined in
+// rules. Returns the Rule that matched, or nil if no match was found.
+func matchImage(image string, rules []api.ImageQualifyRule) *api.ImageQualifyRule {
+	for i := range rules {
+		if ok, _ := path.Match(rules[i].Pattern, image); ok {
```

Comment:
not sure we want to support all the possible patterns `path.Match` allows... can we limit the charset of the pattern to characters that are allowed in image specs (something like `a-zA-Z0-9_-/:@`, I think), and `*`?

Reply:
Do we want to support wildcards in the digest part?

I've looked at this in a number of ways:

- take your suggested character set, and add "*", anything outside of that is bogus, though therre is the perennial risk that image specifications chnage underneath our feet
- look at explicitly referencing these [regexps](https://github.com/openshift/origin/blob/master/pkg/image/apis/image/internal/reference/regexp.go), but a) they're internal and b) was curious as to why it's a copy

If we don't want to support wildcards in digests then (I think) an alternative pattern validation scheme may be:  

```
    tmp := strings.Replace(rule.Pattern, "*", "a", -1))
    if _, _, _, err := parsers.ParseImageName(tmp); err != nil {
        // pattern is invalid.
    }
```
I had chosen to be liberal with patterns as we always validate the image reference when it gets qualified. It would, I agree, be nice to be aware up-front that the pattern is invalid. Either way, you would have to correct and restart whether we do validation up front or (possibly) much later when a match occurs.

Having attempted some validation I'm wondering if it really makes sense to have something like:

    a/b@sha256:abcd*ff*11*

Part of me says I would prefer to fix my patterns vis-a-vis baking policy in and then having to wait for a binary release should things go south.

Reply:
> therre is the perennial risk that image specifications chnage underneath our feet

pretty sure we'd have other updates to make in that case as well. I'd rather start as limited as possible to address known use cases and expand further if needed.